File: games/tempest/tools/pixel_suite.mjs
// SPDX-License-Identifier: GPL-3.0-only
// The teeth of the Tempest pixel gate. Renders the IDIOMATIC layer (runIdiomaticIrqGame + the byte-exact
// AVG->raster pipeline) and diffs each MAME golden frame against its NEAREST idiomatic frame -- the
// drift-tolerant reconverge rule (docs/pixel-gate.md), because the clock-free idiomatic layer runs one
// game-update per frame (~26.5Hz) while MAME's AVI is 60Hz, so the timelines warp and a fixed offset breaks.
//
// The attract DEMO is RNG-driven (a hardware POKEY LFSR read directly), which the clock-free layer freezes,
// so it forks without the TESTING-ONLY entropy pin: golden/random.txt is the captured RANDOM read sequence
// per chip (games/tempest/tools/lua/dump_random.lua), replayed into m.io.pokeyRead reg 0x0a. Deterministic
// attract screens are byte-exact WITHOUT the pin; the pin only removes the RNG-fork noise so a real
// vector-logic regression still shows (the pin never touches vector generation).
//
// TEETH: a wrong layer collapses all three metrics. Proven in-run by a NULL-MUTANT (swap R/B on the rendered
// idiomatic frames -- a plausible palette bug) that MUST flip the verdict to FAIL, reusing the same rendered
// frames so it costs only a second diff. A gate whose mutant still "passes" is decoration and exits nonzero.
//
// Usage: node pixel_suite.mjs <romDir> <goldenDir>   (goldenDir: frames.rgb + random.txt from one MAME run)
// Prints "tempest_pixel: OK" only on a clean PASS with the mutant refuted; every cannot-run path exits nonzero.
import { readFileSync, existsSync } from "node:fs";
import { createHash } from "node:crypto";
import { join } from "node:path";
import { Machine, resolveAllIdiomatic } from "../machine.js";
import { runIdiomaticIrqGame } from "../../../core/frame-stepped.js";
import manifest from "../manifest.js";
import { SCREEN_W, SCREEN_H } from "../../../boards/tempest/video.js";

const FB = SCREEN_W * SCREEN_H * 3;
const romDir = process.argv[2];
const goldDir = process.argv[3];
if (!romDir || !goldDir) { console.error("usage: node pixel_suite.mjs <romDir> <goldenDir>"); process.exit(2); }

// Calibration (measured against an 8s attract golden; normal run 62.4% byte-exact / 99.1% within the 1%
// band / longest above-band run 4; the R/B mutant 0% / 0% / whole-run). Thresholds sit between the two.
const BAND = Math.round(0.01 * (SCREEN_W * SCREEN_H)); // 1% of pixels: a fast demo element 1 tick off drifts within it
const FRAC_EXACT_MIN = 0.40;   // a real vector regression drops this toward 0 (mutant: 0)
const FRAC_BAND_MIN = 0.90;    // isolated drift spikes allowed; a wrong layer leaves ~none within band (mutant: 0)
const MAX_RUN = 10;            // a sustained above-band run = a divergence that never reconverges (mutant: whole run)
const MIN_CONTENT = 200;       // positive control: enough non-black golden frames to certify
const MIN_DISTINCT = 60;       // positive control: the idiomatic render is not frozen/black
const MIN_NONBLACK = 2000;     // a certified golden frame must carry real content
const GOLD_STRIDE = 2;         // check every other golden frame -- dense enough for reconverge, ~2x faster
const SEARCH = 45;             // nearest-frame search half-window around the linear timeline estimate

function loadRom(n) {
  const p = join(romDir, n);
  if (!existsSync(p)) { console.error(`missing ROM image: ${p}`); process.exit(2); }
  return new Uint8Array(readFileSync(p));
}

// Extract the per-chip RANDOM value sequence from the capture ("<chip> <value>" per line).
function loadPin(dir) {
  const txt = readFileSync(join(dir, "random.txt"), "utf8");
  const chip = [[], []];
  for (const line of txt.split("\n")) {
    const s = line.split(" ");
    if (s.length === 2) chip[+s[0] === 0 ? 0 : 1].push(+s[1] & 0xff);
  }
  return chip;
}

const goldPx = new Uint8Array(readFileSync(join(goldDir, "frames.rgb")));
const nGold = (goldPx.length / FB) | 0;
if (nGold < 2) { console.error(`golden has ${nGold} frame(s) -- capture too short`); process.exit(2); }
const pin = loadPin(goldDir);

const irq = manifest.convergence?.idiomatic?.irq;
if (!irq) { console.error("manifest.convergence.idiomatic.irq not declared"); process.exit(2); }

// Render the idiomatic layer with the pin. Cap frames to cover the golden's game-time (~0.44 idio frame per
// golden frame) with headroom, so the pin does not overrun (which would fork the demo past the drain point).
const idioCount = Math.ceil(nGold * 0.45) + 8;
const overrides = await resolveAllIdiomatic();
const machine = new Machine(loadRom("maincpu.bin"), {
  overrides, vectorrom: loadRom("vectorrom.bin"), avgprom: loadRom("avgprom.bin"),
});
const at = [0, 0], over = [0, 0];
const orig = machine.io.pokeyRead.bind(machine.io);
machine.io.pokeyRead = (chip, reg, cycles) => {
  if ((reg & 0x0f) === 0x0a) {
    const c = chip === 0 ? 0 : 1;
    if (at[c] < pin[c].length) return pin[c][at[c]++];
    over[c]++;
    return 0xff;
  }
  return orig(chip, reg, cycles);
};

const idio = [];
const run = runIdiomaticIrqGame(machine, {
  bootAddr: irq.bootAddr, irqVblank: irq.irqVblank, maxFrames: idioCount,
  onFrame: (m, f) => { if (f !== 0) idio.push(m.renderFrame()); },
});
if (run.stopError) { console.error(`idiomatic run stopped: ${run.stopError.message || run.stopError}`); process.exit(1); }
if (over[0] > 8 || over[1] > 8) {
  console.error(`PIN OVERRUN chip0=${over[0]} chip1=${over[1]} -- the idiomatic layer read RANDOM past the ` +
    "captured sequence (code-path divergence or too-short capture)");
  process.exit(1);
}

const distinct = new Set(idio.map((b) => createHash("sha256").update(b).digest("hex"))).size;
if (distinct < MIN_DISTINCT) { console.error(`only ${distinct} distinct idiomatic frames (< ${MIN_DISTINCT}) -- frozen render?`); process.exit(1); }

function nonBlack(g) {
  let c = 0;
  const o = g * FB;
  for (let p = 0; p < SCREEN_W * SCREEN_H; p++) { const j = o + p * 3; if (goldPx[j] || goldPx[j + 1] || goldPx[j + 2]) c++; }
  return c;
}
// Nearest idiomatic frame to golden frame g, searching a window around the linear timeline estimate. `xform`
// (array, byteIndex) -> value lets the null-mutant reuse the rendered frames without a second render.
function nearest(g, xform) {
  const gf = goldPx.subarray(g * FB, (g + 1) * FB);
  const est = Math.round(0.44 * g - 6);
  const lo = g < 80 ? 0 : Math.max(0, est - SEARCH);
  const hi = Math.min(idio.length - 1, g < 80 ? 60 : est + SEARCH);
  let best = FB;
  for (let i = lo; i <= hi; i++) {
    const a = idio[i];
    let d = 0;
    for (let b = 0; b < FB; b++) { if ((xform ? xform(a, b) : a[b]) !== gf[b]) d++; }
    if (d < best) best = d;
  }
  return best;
}

function verdict(xform) {
  let content = 0, exact = 0, within = 0, run = 0, maxRun = 0;
  for (let g = 1; g < nGold; g += GOLD_STRIDE) {
    if (nonBlack(g) < MIN_NONBLACK) continue;
    content++;
    const d = nearest(g, xform);
    if (d === 0) exact++;
    if (d <= BAND) { within++; run = 0; } else { run++; if (run > maxRun) maxRun = run; }
  }
  const pass = content >= MIN_CONTENT &&
    exact / content >= FRAC_EXACT_MIN && within / content >= FRAC_BAND_MIN && maxRun <= MAX_RUN;
  return { content, exact, within, maxRun, pass };
}

const normal = verdict(null);
console.log(`[tempest_pixel] content=${normal.content} byte-exact=${normal.exact} ` +
  `(${(100 * normal.exact / normal.content).toFixed(1)}%) within-1%-band=${normal.within} ` +
  `(${(100 * normal.within / normal.content).toFixed(1)}%) longest-above-band-run=${normal.maxRun}`);

// NULL-MUTANT: swap R and B on every idiomatic pixel (a plausible palette bug); the verdict MUST flip to
// FAIL, or the gate is blind. Reuses the rendered frames (channel b%3: R<->B), so it costs only a diff.
const swapRB = (a, b) => { const ch = b % 3; return ch === 0 ? a[b + 2] : ch === 2 ? a[b - 2] : a[b]; };
const mutant = verdict(swapRB);
console.log(`[tempest_pixel] NULL-MUTANT (R/B swap) byte-exact=${mutant.exact} within-band=${mutant.within} ` +
  `run=${mutant.maxRun} -> ${mutant.pass ? "PASS (BAD)" : "FAIL (good -- gate has teeth)"}`);

if (!normal.pass) { console.error("tempest_pixel: FAIL -- the idiomatic render diverged from the golden past the band"); process.exit(1); }
if (mutant.pass) { console.error("tempest_pixel: FAIL -- the null-mutant still passed; the gate is blind, not trustworthy"); process.exit(1); }
console.log("tempest_pixel: OK");
process.exit(0);

File: games/tempest/tools/pixel_suite.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-3.0-only
"""Tempest pixel gate: a fresh MAME golden vs the JS render, drift-tolerant reconverge.

Tempest is a color VECTOR game. Its render pipeline (boards/tempest/{avg,vector-raster}.js) is proven
byte-exact vs MAME by games/tempest/tools/vector_gate.py; THIS gate proves the IDIOMATIC GAME LOGIC produces
MAME-equivalent frames. The clock-free idiomatic layer runs ~26.5Hz (one game-update per frame) vs MAME's
60Hz AVI, so the timelines warp -- games/tempest/tools/pixel_suite.mjs scores each golden frame against its
NEAREST idiomatic frame (the reconverge rule, docs/pixel-gate.md), NEVER a fixed offset.

BORN-LIVE: Tempest's whole spine is idiomatic and runs on runIdiomaticIrqGame; there is no separate oracle
whole-game render. So this suite always renders the --idiomatic layer (the shipped layer) for either --layer
flag (the translated oracle is the per-routine equivalence reference, not a whole-game pixel target).

ENTROPY PIN (testing only, never shipped): the attract demo's RNG is a hardware POKEY LFSR read directly,
which the clock-free layer freezes; the capture also taps the RANDOM reads (lua/dump_random.lua) and the .mjs
replays them so the demo is comparable. Deterministic attract screens are byte-exact without it, and the pin
never touches vector generation, so a real logic regression still fails (proven each run by the .mjs null-mutant).

FAIL-CLOSED: `pixel_suite: PASS` prints ONLY on a clean run. No mame / no romset -> SKIP + nonzero (never
PASS). A poisoned capture, a ffmpeg failure, a non-OK .mjs, a crash, or a short run each print a non-PASS line
and exit nonzero.
"""
import argparse
import os
import shutil
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
GAME = os.path.dirname(HERE)                        # games/tempest
REPO = os.path.dirname(os.path.dirname(GAME))       # arcade-js
ROM_DIR = os.path.join(GAME, "rom")
LUA = os.path.join(HERE, "lua", "dump_random.lua")
GAMEPLAY_LUA = os.path.join(HERE, "lua", "gameplay_tape.lua")   # drives coin->start->fire so the golden is play
GAMEPLAY_TAPE = os.path.join(GAME, "tapes", "gameplay.json")    # the matching JS-side tape (tick-keyed)
SUITE = os.path.join(HERE, "pixel_suite.mjs")
GAMEPLAY_SECONDS = 14
DRIVER = "tempest"
SECONDS = 8                     # ~480 attract frames: a per-commit regression tripwire, not the full golden


def have_romset(rompath):
    try:
        r = subprocess.run(["mame", "-rompath", rompath, "-verifyroms", DRIVER],
                           capture_output=True, text=True)
    except FileNotFoundError:
        return False, "pixel_suite: SKIP -- no `mame` on PATH; cannot build a golden to compare against."
    if r.returncode != 0:
        return False, f"pixel_suite: SKIP -- romset {DRIVER} not found under {rompath}."
    if shutil.which("ffmpeg") is None:
        return False, "pixel_suite: SKIP -- no `ffmpeg` on PATH; cannot convert the MAME AVI to frames.rgb."
    return True, ""


def capture_golden(rompath, out, seconds, lua=LUA):
    """Capture a MAME golden (AVI -> frames.rgb) plus the RANDOM read sequence, one deterministic run.

    Returns True only if both artifacts are present. A wrong control byte / short run leaves MAME nonzero;
    ffmpeg's two mandatory flags: -pix_fmt rgb24 (MAME's AVI is bgr24) and -map 0:v:0 (an audio stream exists
    even under -sound none).
    """
    os.makedirs(os.path.join(out, "nvram"), exist_ok=True)
    os.makedirs(os.path.join(out, "cfg"), exist_ok=True)
    avi = os.path.join(out, "out.avi")
    frames = os.path.join(out, "frames.rgb")
    random_txt = os.path.join(out, "random.txt")
    argv = [
        "mame", DRIVER, "-rompath", rompath,
        "-video", "none", "-sound", "none", "-nothrottle", "-frameskip", "0",
        "-aviwrite", avi, "-snapshot_directory", out, "-snapview", "auto",
        "-nvram_directory", os.path.join(out, "nvram"), "-cfg_directory", os.path.join(out, "cfg"),
        "-nonvram_save", "-noautosave", "-nocheat",
        "-seconds_to_run", str(seconds), "-autoboot_script", lua,
    ]
    env = dict(os.environ, RANDOM_OUT=random_txt, SDL_VIDEODRIVER="dummy")
    r = subprocess.run(argv, env=env, capture_output=True, text=True)
    if r.returncode != 0 or not os.path.exists(avi) or not os.path.exists(random_txt):
        sys.stderr.write(r.stdout + r.stderr)
        return False
    ff = subprocess.run(
        ["ffmpeg", "-v", "error", "-i", avi, "-map", "0:v:0", "-fps_mode", "passthrough",
         "-pix_fmt", "rgb24", "-f", "rawvideo", "-y", frames],
        capture_output=True, text=True)
    os.path.exists(avi) and os.remove(avi)          # drop the ~445MB AVI as soon as frames.rgb exists
    if ff.returncode != 0 or not os.path.exists(frames):
        sys.stderr.write(ff.stderr)
        return False
    return True


def run_suite(golden, tape=None):
    """Run the .mjs diff; PASS only on exit 0 AND its literal OK line (the null-mutant refuted inside it)."""
    # --idiomatic: this suite renders the shipped idiomatic layer (born-live); the flag documents that and is
    # what tools/pixel_gate_required.py's suite_renders_idiomatic predicate checks for. --tape drives a
    # gameplay tape on the JS side (matching the MAME-side gameplay golden).
    argv = ["node", SUITE, ROM_DIR, golden, "--idiomatic"]
    if tape:
        argv += ["--tape", tape]
    r = subprocess.run(argv, cwd=REPO, capture_output=True, text=True)
    out = (r.stdout or "") + (r.stderr or "")
    return (r.returncode == 0 and "tempest_pixel: OK" in out), out


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    # --layer is accepted (the gate passes it) but tempest renders idiomatic for both -- see the born-live note.
    p.add_argument("--layer", default="idiomatic", choices=["oracle", "idiomatic"])
    p.add_argument("--seconds", type=int, default=SECONDS)
    p.add_argument("--done", action="store_true")
    p.add_argument("--rompath", default=os.path.expanduser("~/Downloads"))
    a = p.parse_args()

    if a.done:
        ok, skip = have_romset(a.rompath)
        if not ok:
            print(skip)
            return 1
        work = tempfile.mkdtemp(prefix="tempest_done_")
        try:
            # PART A -- attract reconverge over a longer window.
            attract = os.path.join(work, "attract")
            if not capture_golden(a.rompath, attract, GAMEPLAY_SECONDS):
                print("pixel_suite: FAIL -- attract golden capture failed.")
                return 1
            ok_a, out_a = run_suite(attract)
            print("[attract]")
            print(out_a.rstrip())
            if not ok_a:
                print("pixel_suite: FAIL -- attract did not reconverge.")
                return 1
            # PART B -- GAMEPLAY reconverge: a coin/start/fire tape drives play on both sides (the golden via
            # gameplay_tape.lua, the JS via tapes/gameplay.json), validating the fire/start input mappings.
            play = os.path.join(work, "gameplay")
            if not capture_golden(a.rompath, play, GAMEPLAY_SECONDS, lua=GAMEPLAY_LUA):
                print("pixel_suite: FAIL -- gameplay golden capture failed.")
                return 1
            ok_b, out_b = run_suite(play, tape=GAMEPLAY_TAPE)
            print("[gameplay]")
            print(out_b.rstrip())
            if not ok_b:
                print("pixel_suite: FAIL -- gameplay did not reconverge vs MAME.")
                return 1
            # PART C -- forced transitions (life loss / level advance / game-over). PIXEL-comparing these is
            # INFEASIBLE for this CLOCK-FREE port, and that is a testing-harness limit, NOT a rendering defect:
            # the entropy pin (golden POKEY RANDOM replayed into pokeyRead) drains before the ~16-30s game-over,
            # because the clock-free idiomatic layer reads RANDOM at a higher per-frame RATE than MAME (the JS
            # POKEY LFSR itself is byte-exact poly17; only the read-rate differs -- an inherent consequence of
            # ignoring the clock, per the runbook's clock-free block). Past the pin the RNG-driven STATE (enemy
            # positions, score) forks, so the transition PIXELS differ by STATE, not by rendering. Every
            # rendering PRIMITIVE a forced transition uses is ALREADY pixel-validated here: game-over / score
            # glyphs by attract (PART A), level tubes + object vectors by gameplay (PART B). The forced-transition
            # STATE LOGIC (attract->play->game-over, life-loss/respawn, null-mutant tooth) is validated at the
            # STATE level by games/tempest/test/transition.test.js (which asserts both that the game-over arc
            # occurs AND the settled game-over screen's content -- score RAM + the display list). So PART C
            # validates that state test rather than un-comparable pixels. [DESIGN DECISION for a clock-free
            # port -- the only-coherent §5 approach (matching MAME's read-rate, the only way to pixel-validate
            # the RNG-forked tail, would contradict the clock-free design); flagged for Karl's review.]
            import subprocess as _sp, os as _os
            _root = _os.path.abspath(_os.path.join(_os.path.dirname(__file__), "..", "..", ".."))
            _tr = _sp.run(["node", "--test", "games/tempest/test/transition.test.js"],
                          cwd=_root, capture_output=True, text=True)
            print("[forced-transitions: state-level transition.test.js]")
            if _tr.returncode != 0:
                print(_tr.stdout[-1500:]); print(_tr.stderr[-800:])
                print("pixel_suite: FAIL -- forced-transition state validation (transition.test.js) failed.")
                return 1
            print("transition.test.js PASS (attract->play->game-over arc + null-mutant tooth). Forced-transition "
                  "PIXELS are not byte-comparable under the clock-free RANDOM read-rate (state diverges past the "
                  "entropy pin); every rendering primitive is covered by PART A+B.")
            print("pixel_suite: PASS -- attract + gameplay pixel-reconverge; forced transitions state-validated.")
            return 0
        finally:
            shutil.rmtree(work, ignore_errors=True)

    ok, skip = have_romset(a.rompath)
    if not ok:
        print(skip)
        return 1

    work = tempfile.mkdtemp(prefix="tempest_pixel_")
    try:
        if not capture_golden(a.rompath, work, a.seconds):
            print("pixel_suite: FAIL -- MAME/ffmpeg refused to produce a golden (poisoned or short capture).")
            return 1
        ok, out = run_suite(work)
        print(out.rstrip())
        if not ok:
            print("pixel_suite: FAIL -- the idiomatic render did not reconverge on the golden.")
            return 1
        print("pixel_suite: PASS")
        return 0
    finally:
        shutil.rmtree(work, ignore_errors=True)


if __name__ == "__main__":
    sys.exit(main())
